app/src/behavior.c

```diff
@@ -39,6 +41,144 @@ const struct device *z_impl_behavior_get_binding(const char *name) {
     return NULL;
 }
 
+#if IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
+
+int zmk_behavior_get_empty_param_metadata(const struct device *dev,
+                                          struct behavior_parameter_metadata *metadata) {
+    metadata->sets_len = 0;
+    return 0;
+}
+
+#endif // IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
+
+#if IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
+static int validate_hid_usage(uint16_t usage_page, uint16_t usage_id) {
+    LOG_DBG("Validate usage %d in page %d", usage_id, usage_page);
+    switch (usage_page) {
+    case HID_USAGE_KEY:
+        if (usage_id == 0 || (usage_id > ZMK_HID_KEYBOARD_NKRO_MAX_USAGE &&
+                              usage_id < LEFT_CONTROL && usage_id > RIGHT_GUI)) {
+            return -EINVAL;
+        }
+        break;
+    case HID_USAGE_CONSUMER:
+        if (usage_id >
+            COND_CODE_1(IS_ENABLED(CONFIG_ZMK_HID_CONSUMER_REPORT_USAGES_BASIC), (0xFF), (0xFFF))) {
+            return -EINVAL;
+        }
+        break;
+    default:
+        LOG_WRN("Unsupported HID usage page %d", usage_page);
+        return -EINVAL;
+    }
+
+    return 0;
+}
+
+int zmk_behavior_validate_param_values(const struct behavior_parameter_value_metadata *values,
+                                       size_t values_len, uint32_t param) {
+    bool found_match = false;
+
+    if (values_len == 0) {
+        return -ENODEV;
+    }
+
+    for (int v = 0; v < values_len; v++) {
+        const struct behavior_parameter_value_metadata *value_meta = &values[v];
+
+        switch (value_meta->type) {
+        case BEHAVIOR_PARAMETER_VALUE_TYPE_NIL:
+            if (param == 0) {
+                found_match = true;
+            }
+            break;
+        case BEHAVIOR_PARAMETER_VALUE_TYPE_HID_USAGE:
+            if (validate_hid_usage(ZMK_HID_USAGE_PAGE(param), ZMK_HID_USAGE_ID(param)) >= 0) {
+                found_match = true;
+            }
+
+            break;
+        case BEHAVIOR_PARAMETER_VALUE_TYPE_LAYER_INDEX:
+            if (param >= 0 && param < ZMK_KEYMAP_LEN) {
+                found_match = true;
+            }
+            break;
+            /* TODO: Restore with HSV -> RGB refactor
+            case BEHAVIOR_PARAMETER_STANDARD_DOMAIN_HSV:
+                // TODO: No real way to validate? Maybe max brightness?
+                break;
+            */
+        case BEHAVIOR_PARAMETER_VALUE_TYPE_VALUE:
+            if (param == value_meta->value) {
+                found_match = true;
+            }
+            break;
+        case BEHAVIOR_PARAMETER_VALUE_TYPE_RANGE:
+            if (param >= value_meta->range.min && param <= value_meta->range.max) {
+                found_match = true;
+            }
+            break;
+        }
+
+        if (found_match) {
+            break;
+        }
+    }
+
+    return found_match ? 0 : -ENOTSUP;
```

Comment:
Any reason this one return -ENOTSUP while most of the others return -EINVAL on a failure to match?

Reply:
Oversight. Corrected. Thanks!